kornia/geometry/liegroup/se2.py

```diff
@@ -0,0 +1,252 @@
+# kornia.geometry.se2 module inspired by Sophus-sympy.
+# https://github.com/strasdat/Sophus/blob/master/sympy/sophus/se2.py
+from typing import Optional, Tuple, Union
+
+from kornia.core import Module, Parameter, Tensor, concatenate, pad, rand, stack, tensor, where
+from kornia.geometry.liegroup._utils import check_se2_r_t_shape, check_v_shape
+from kornia.geometry.liegroup.so2 import So2
+from kornia.testing import KORNIA_CHECK, KORNIA_CHECK_TYPE
+
+
+class Se2(Module):
+    r"""Base class to represent the Se2 group.
+
+    The SE(2) is the group of rigid body transformations about the origin of two-dimensional Euclidean
+    space :math:`R^2` under the operation of composition.
+    See more:
+
+    Example:
+        >>> so2 = So2.identity(1)
+        >>> t = torch.ones((1, 2))
+        >>> se2 = Se2(so2, t)
+        >>> se2
+        rotation: Parameter containing:
+        tensor([1.+0.j], requires_grad=True)
+        translation: Parameter containing:
+        tensor([[1., 1.]], requires_grad=True)
+    """
+
+    def __init__(self, rotation: So2, translation: Tensor) -> None:
+        """Constructor for the base class.
+
+        Internally represented by a complex number `z` and a translation 2-vector.
+
+        Args:
+            r: So2 group encompassing a rotation.
+            t: translation vector with the shape of :math:`(B, 2)`.
+
+        Example:
+            >>> so2 = So2.identity(1)
+            >>> t = torch.ones((1, 2))
+            >>> se2 = Se2(so2, t)
+            >>> se2
+            rotation: Parameter containing:
+            tensor([1.+0.j], requires_grad=True)
+            translation: Parameter containing:
+            tensor([[1., 1.]], requires_grad=True)
+        """
+        super().__init__()
+        KORNIA_CHECK_TYPE(rotation, So2)
+        # TODO change to KORNIA_CHECK_SHAPE once there is multiple shape support
+        check_se2_r_t_shape(rotation, translation)
+        self._rotation = rotation
+        self._translation = Parameter(translation)
+
+    def __repr__(self) -> str:
+        return f"rotation: {self.r}\ntranslation: {self.t}"
+
+    def __getitem__(self, idx) -> 'Se2':
+        return Se2(self._rotation[idx], self._translation[idx][None])
+
+    def __mul__(self, right: "Se2") -> Union['Se2', Tensor]:
+        """Compose two Se2 transformations.
+
+        Args:
+            right: the other Se2 transformation.
+
+        Return:
+            The resulting Se2 transformation.
+        """
+        out: Union['Se2', Tensor]
+        if isinstance(right, Se2):
+            KORNIA_CHECK_TYPE(right, Se2)
+            so2 = self.so2
```

Comment:
```suggestion
            so2L So2 = self.so2
```
maybe this will disambiguate ? /cc @johnnv1

Reply:
Isn't the typing of so2 and se2 mixed up?

Reply:
Maybe can do this here? 


```diff
--- a/kornia/geometry/liegroup/se2.py
+++ b/kornia/geometry/liegroup/se2.py
@@ -1,6 +1,6 @@
 # kornia.geometry.se2 module inspired by Sophus-sympy.
 # https://github.com/strasdat/Sophus/blob/master/sympy/sophus/se2.py
-from typing import Optional, Tuple, Union
+from typing import Optional, Tuple, overload
 
 from kornia.core import Module, Parameter, Tensor, concatenate, pad, rand, stack, tensor, where
 from kornia.geometry.liegroup._utils import check_se2_r_t_shape, check_v_shape
@@ -58,7 +58,15 @@ class Se2(Module):
     def __getitem__(self, idx) -> 'Se2':
         return Se2(self._rotation[idx], self._translation[idx][None])
 
-    def __mul__(self, right: "Se2") -> Union['Se2', Tensor]:
+    @overload
+    def __mul__(self, right: 'Se2') -> 'Se2':
+        ...
+
+    @overload
+    def __mul__(self, right: Tensor) -> Tensor:
+        ...
+
+    def __mul__(self, right):
         """Compose two Se2 transformations.
 
         Args:
@@ -67,21 +75,18 @@ class Se2(Module):
         Return:
             The resulting Se2 transformation.
         """
-        out: Union['Se2', Tensor]
         if isinstance(right, Se2):
             KORNIA_CHECK_TYPE(right, Se2)
-            so2 = self.so2
-            r: So2 = so2 * right.so2
-            t: Tensor = self.t + so2 * right.t
-            out = Se2(r, t)
+            r = self.so2 * right.so2
+            t = self._translation + self.so2 * right._translation
+            return Se2(r, t)
         elif isinstance(right, Tensor):
             KORNIA_CHECK_TYPE(right, Tensor)
             # TODO change to KORNIA_CHECK_SHAPE once there is multiple shape support
             check_se2_r_t_shape(self.so2, right)
-            out = self.so2 * right + self.t
+            return self.so2 * right + self.t
         else:
             raise TypeError(f"Unsupported type: {type(right)}")
-        return out
 
     @property
     def so2(self) -> So2:
diff --git a/kornia/geometry/liegroup/so2.py b/kornia/geometry/liegroup/so2.py
index 20b7aff3..7658e063 100644
--- a/kornia/geometry/liegroup/so2.py
+++ b/kornia/geometry/liegroup/so2.py
@@ -1,6 +1,6 @@
 # kornia.geometry.so2 module inspired by Sophus-sympy.
 # https://github.com/strasdat/Sophus/blob/master/sympy/sophus/so2.py
-from typing import Optional, Union
+from typing import Optional, overload
 
 from kornia.core import Module, Parameter, Tensor, complex, rand, stack, tensor, zeros_like
 from kornia.geometry.liegroup._utils import (
@@ -56,7 +56,15 @@ class So2(Module):
     def __getitem__(self, idx: int) -> 'So2':
         return So2(self._z[idx][..., None])
 
-    def __mul__(self, right: Union['So2', Tensor]) -> Union['So2', Tensor]:
+    @overload
+    def __mul__(self, right: 'So2') -> 'So2':
+        ...
+
+    @overload
+    def __mul__(self, right: Tensor) -> Tensor:
+        ...
+
+    def __mul__(self, right):
         """Performs a left-multiplication either rotation concatenation or point-transform.
 
         Args:
@@ -65,9 +73,8 @@ class So2(Module):
         Return:
             The resulting So2 transformation.
         """
-        out: Union['So2', Tensor]
         if isinstance(right, So2):
-            out = So2(self.z * right.z)
+            return So2(self.z * right.z)
         elif isinstance(right, Tensor):
             # TODO change to KORNIA_CHECK_SHAPE once there is multiple shape support
             check_so2_t_shape(right)
@@ -75,10 +82,9 @@ class So2(Module):
             y = right[..., 1]
             real = self.z.real
             imag = self.z.imag
-            out = stack((real * x - imag * y, imag * x + real * y), -1)
+            return stack((real * x - imag * y, imag * x + real * y), -1)
         else:
             raise TypeError(f"Not So2 or Tensor type. Got: {type(right)}")
-        return out
```

Reply:
sounds good to me

Reply:
nice! Thanks @johnnv1